File: contrib/phpcr/PHPCR/nodetype/NodeType.php
<?php

/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  The ASF licenses this file to You
 * under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


require_once 'PHPCR/Value.php';
require_once 'PHPCR/nodetype/PropertyDefinition.php';
require_once 'PHPCR/nodetype/NodeDefinition.php';
require_once 'PHPCR/nodetype/NodeType.php';


/**
 * Represents a node type.
 *
 * @author Markus Nix <mnix@mayflower.de>
 * @package phpcr
 * @subpackage nodetype
 */
interface NodeType
{
    /**
     * Returns the name of the node type.
     *
     * @return the name of the node type
     */
    public function getName();

    /**
     * Returns <code>true</code> if this node type is a mixin node type.
     * Returns <code>false</code> if this node type is a primary node type.
     *
     * @return a boolean
     */
    public function isMixin();

    /**
     * Returns <code>true</code> if nodes of this type must support orderable child nodes; returns <code>false</code>
     * otherwise. If a node type returns <code>true</code> on a call to this method, then all nodes of that node type
     * <i>must</i> support the method {@link Node#orderBefore}. If a node type returns <code>false</code>
     * on a call to this method, then nodes of that node type <i>may</i> support these ordering methods. Only the primary
     * node type of a node controls that node's status in this regard. This setting on a mixin node type will not have any effect
     * on the node.
     *
     * @return Returns <code>true</code> if nodes of this type must support orderable child nodes; returns
     * <code>false</code> otherwise.
     */
    public function hasOrderableChildNodes();

    /**
     * Returns the name of the primary item (one of the child items of the node's of this node type).
     * If this node has no primary item, then this method returns <code>null</code>.
     * This indicator is used by the method {@link Node#getPrimaryItem()}.
     *
     * @return the name of the primary item.
     */
    public function getPrimaryItemName();

    /**
     * Returns all supertypes of this node type including both those directly
     * declared and those inherited. For primary types, this list will always
     * include at least <code>nt:base</code>. For mixin types, there is no
     * required base type.
     *
     * @see #getDeclaredSupertypes
     *
     * @return an array of <code>NodeType</code> objects.
     */
    public function getSupertypes();

    /**
     * Returns all <i>direct</i> supertypes as specified in the declaration of
     * <i>this</i> node type. In single inheritance systems this will always be
     * an array of size 0 or 1. In systems that support multiple inheritance of
     * node types this array may be of size greater than 1.
     *
     * @see #getSupertypes
     *
     * @return an array of <code>NodeType</code> objects.
     */
    public function getDeclaredSupertypes();

    /**
     * Returns true if this node type is <code>nodeTypeName</code>
     * or a subtype of <code>nodeTypeName</code>, otherwise returns
     * <code>false</code>.
     * @param nodeTypeName the name of a node type.
     * @return a boolean
     */
    public function isNodeType( $nodeTypeName );

    /**
     * Returns an array containing the property definitions of this node type,
     * including the property definitions inherited from supertypes of this node
     * type.
     *
     * @see #getDeclaredPropertyDefinitions
     *
     * @return an array containing the property definitions.
     */
    public function getPropertyDefinitions();

    /**
     * Returns an array containing the property definitions explicitly specified
     * in the declaration of <i>this</i> node type. This does <i>not</i> include
     * property definitions inherited from supertypes of this node type.
     *
     * @see #getPropertyDefinitions
     *
     * @return an array containing the property definitions.
     */
    public function getDeclaredPropertyDefinitions();

    /**
     * Returns an array containing the child node definitions of this node type,
     * including the child node definitions inherited from supertypes of this
     * node type.
     *
     * @see #getDeclaredChildNodeDefinitions
     *
     * @return an array containing the child node definitions.
     */
    public function getChildNodeDefinitions();

    /**
     * Returns an array containing the child node definitions explicitly
     * specified in the declaration of <i>this</i> node type. This does
     * <i>not</i> include child node definitions inherited from supertypes of
     * this node type.
     *
     * @see #getChildNodeDefinitions
     * @return an array containing the child node definitions.
     */
    public function getDeclaredChildNodeDefinitions();

    /**
     * Returns <code>true</code> if setting <code>propertyName</code> to
     * <code>value</code> is allowed by this node type. Otherwise returns
     * <code>false</code>.
     *
     * @param propertyName The name of the property
     * @param value A <code>Value</code> object.
     */
    public function canSetProperty( $propertyName, $value );

    /**
     * Returns <code>true</code> if adding a child node called
     * <code>childNodeName</code> is allowed by this node type.
     * <p>
     *
     * @param childNodeName The name of the child node.
     * @param nodeTypeName The name of the node type of the child node.
     */
    public function canAddChildNode( $childNodeName, $nodeTypeName = null );

    /**
     * Returns true if removing the child item called <code>itemName</code> is allowed by this node type.
     * Otherwise returns <code>false</code>.
     *
     * @param itemName The name of the child item
     */
    public function canRemoveItem( $itemName );
}

?>

File: contrib/jackrabbit-ntdoc/src/main/java/org/apache/jackrabbit/ntdoc/model/NodeType.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.jackrabbit.ntdoc.model;

import java.util.*;

/**
 * This class implements the node type.
 */
public final class NodeType
        implements Comparable {
    /**
     * Name of type.
     */
    private final String name;

    /**
     * Namespace.
     */
    private String namespace;

    /**
     * Super types.
     */
    private final HashSet superTypes;

    /**
     * Mixin.
     */
    private boolean mixin;

    /**
     * Orderable.
     */
    private boolean orderable;

    /**
     * List of child items.
     */
    private final List itemDefs;

    /**
     * Node type set.
     */
    private NodeTypeSet nodeTypeSet;

    /**
     * Construct the node type.
     */
    public NodeType(String name) {
        this.name = name;
        this.itemDefs = new ArrayList();
        this.superTypes = new HashSet();
    }

    /**
     * Return the name.
     */
    public String getName() {
        return this.name;
    }

    /**
     * Return the namespace URI.
     */
    public String getNamespace() {
        return this.namespace != null ? this.namespace : "";
    }

    /**
     * Set the namespace URI.
     */
    public void setNamespace(String namespace) {
        this.namespace = namespace;
    }

    /**
     * Return the super types.
     */
    public String[] getSuperTypes() {
        if (!isMixin()) {
            addSuperType("nt:base");
        }

        return (String[]) this.superTypes.toArray(new String[this.superTypes.size()]);
    }

    /**
     * Set the super types.
     */
    public void setSuperTypes(String[] superTypes) {
        setSuperTypes(Arrays.asList(superTypes));
    }

    /**
     * Set the super types.
     */
    public void setSuperTypes(Collection superTypes) {
        this.superTypes.clear();
        this.superTypes.addAll(superTypes != null ? superTypes : Collections.EMPTY_LIST);
    }

    /**
     * Add a super type.
     */
    public void addSuperType(String superType) {
        this.superTypes.add(superType);
    }

    /**
     * Return true if mixin type.
     */
    public boolean isMixin() {
        return this.mixin;
    }

    /**
     * Set true if mixin type.
     */
    public void setMixin(boolean mixin) {
        this.mixin = mixin;
    }

    /**
     * Return true if it has orderable child nodes.
     */
    public boolean isOrderable() {
        return this.orderable;
    }

    /**
     * Set true if it has orderable child nodes.
     */
    public void setOrderable(boolean orderable) {
        this.orderable = orderable;
    }

    /**
     * Return the primary item name.
     */
    public ItemDef getPrimaryItem() {
        for (Iterator i = this.itemDefs.iterator(); i.hasNext();) {
            ItemDef def = (ItemDef) i.next();
            if (def.isPrimary()) {
                return def;
            }
        }

        return null;
    }

    /**
     * Return the primary item name.
     */
    public String getPrimaryItemName() {
        ItemDef def = getPrimaryItem();
        return def != null ? def.getName() : "";
    }

    /**
     * Add item def.
     */
    public void addItemDef(ItemDef def) {
        this.itemDefs.add(def);
        def.setDeclaringNodeType(this);
    }

    /**
     * Find all inherited nodes.
     */
    public NodeType[] getInheritedTypes() {
        return this.nodeTypeSet.getInheritedTypes(this);
    }

    /**
     * Return item def list.
     */
    private List getItemDefList(boolean properties, boolean nodes) {
        return getItemDefList(this.itemDefs, properties, nodes);
    }

    /**
     * Return inherited item def list.
     */
    private List getInheritedItemDefList(boolean properties, boolean nodes) {
        ArrayList list = new ArrayList();
        NodeType[] types = getInheritedTypes();

        for (int i = 0; i < types.length; i++) {
            list.addAll(Arrays.asList(types[i].getItemDefs()));
        }

        return getItemDefList(list, properties, nodes);
    }

    /**
     * Return item def list.
     */
    private List getItemDefList(List source, boolean properties, boolean nodes) {
        ArrayList list = new ArrayList();
        for (Iterator i = source.iterator(); i.hasNext();) {
            ItemDef def = (ItemDef) i.next();

            if (properties && (def instanceof PropertyDef)) {
                list.add(def);
            }

            if (nodes && (def instanceof NodeDef)) {
                list.add(def);
            }
        }

        Collections.sort(list);
        return list;
    }

    /**
     * Return item defs.
     */
    public ItemDef[] getItemDefs() {
        List list = getItemDefList(true, true);
        return (ItemDef[]) list.toArray(new ItemDef[list.size()]);
    }

    /**
     * Return property defs.
     */
    public PropertyDef[] getPropertyDefs() {
        List list = getItemDefList(true, false);
        return (PropertyDef[]) list.toArray(new PropertyDef[list.size()]);
    }

    /**
     * Return node defs.
     */
    public NodeDef[] getNodeDefs() {
        List list = getItemDefList(false, true);
        return (NodeDef[]) list.toArray(new NodeDef[list.size()]);
    }

    /**
     * Return inherited item defs.
     */
    public ItemDef[] getInheritedItemDefs() {
        List list = getInheritedItemDefList(true, true);
        return (ItemDef[]) list.toArray(new ItemDef[list.size()]);
    }

    /**
     * Return inherited property defs.
     */
    public PropertyDef[] getInheritedPropertyDefs() {
        List list = getInheritedItemDefList(true, false);
        return (PropertyDef[]) list.toArray(new PropertyDef[list.size()]);
    }

    /**
     * Return inherited node defs.
     */
    public NodeDef[] getInheritedNodeDefs() {
        List list = getInheritedItemDefList(false, true);
        return (NodeDef[]) list.toArray(new NodeDef[list.size()]);
    }

    /**
     * Return the node type set.
     */
    public NodeTypeSet getNodeTypeSet() {
        return this.nodeTypeSet;
    }

    /**
     * Set the node type set.
     */
    public void setNodeTypeSet(NodeTypeSet nodeTypeSet) {
        this.nodeTypeSet = nodeTypeSet;
    }

    /**
     * Compare to.
     */
    public int compareTo(Object o) {
        return this.name.compareTo(((NodeType) o).name);
    }
}
